internal/jobs/jobobject.go

```diff
@@ -0,0 +1,376 @@
+		}
+	}
+
+	if limits.cpuRate != 0 || limits.cpuWeight != 0 || limits.cpuCount != 0 {
+		cpuInfo := winapi.JOBOBJECT_CPU_RATE_CONTROL_INFORMATION{}
+		if limits.cpuCount != 0 {
+			numProcessors := uint32(processorinfo.ProcessorCount())
+			if limits.cpuCount > numProcessors {
+				limits.cpuCount = numProcessors
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP
+			cpuInfo.Value = calculateJobCPURate(limits.cpuCount, numProcessors)
+		} else if limits.cpuWeight != 0 {
+			if limits.cpuWeight < processorWeightMin || limits.cpuWeight > processorWeightMax {
+				return fmt.Errorf("processor weight value of `%d` is invalid", limits.cpuWeight)
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_WEIGHT_BASED
+			cpuInfo.Value = calculateJobCPUWeight(limits.cpuWeight)
+		} else if limits.cpuRate != 0 {
+			if limits.cpuRate < processorLimitMin || limits.cpuRate > processorLimitMax {
+				return fmt.Errorf("processor rate of `%d` is invalid", limits.cpuRate)
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP
+			cpuInfo.Value = limits.cpuRate
+		}
+		_, err := windows.SetInformationJobObject(job.jobHandle, windows.JobObjectCpuRateControlInformation, uintptr(unsafe.Pointer(&cpuInfo)), uint32(unsafe.Sizeof(cpuInfo)))
+		if err != nil {
+			return fmt.Errorf("failed to set cpu limit info on job object: %s", err)
+		}
+	}
+
+	if limits.maxBandwidth != 0 || limits.maxIops != 0 {
+		ioInfo := winapi.JOBOBJECT_IO_RATE_CONTROL_INFORMATION{
+			ControlFlags: winapi.JOB_OBJECT_IO_RATE_CONTROL_ENABLE,
+		}
+		if limits.maxBandwidth != 0 {
+			ioInfo.MaxBandwidth = limits.maxBandwidth
+		}
+		if limits.maxIops != 0 {
+			ioInfo.MaxIops = limits.maxIops
+		}
+		_, err := winapi.SetIoRateControlInformationJobObject(job.jobHandle, &ioInfo)
+		if err != nil {
+			return fmt.Errorf("failed to set IO limit info on job object: %s", err)
+		}
+	}
+	return nil
+}
+
+// CreateJobObject creates a job object, attaches an IO completion port to use
+// for notifications and then returns an object with the corresponding handles.
+func CreateJobObject(name string) (*JobObject, error) {
+	jobHandle, err := windows.CreateJobObject(nil, windows.StringToUTF16Ptr(name))
+	if err != nil {
+		return nil, err
+	}
+
+	iocpChan := make(chan IOCPNotif, 10)
+	jobMap[uint32(jobHandle)] = iocpChan
+	// If the IOCP we'll be using to receive messages for all jobs hasn't been
+	// created, create it and start polling.
+	var initIOErr error
+	ioInitOnce.Do(func() {
+		h, err := windows.CreateIoCompletionPort(windows.InvalidHandle, 0, 0, 0xffffffff)
+		// If this fails we don't have any way of monitoring when a job container has
+		// shutdown fully. This is run once via sync.Once.Do so we can't just bubble up an error
+		// here.
+		if err != nil {
+			initIOErr = err
```

Comment:
We should cache `initIOErr` as a global along with `ioCompletionPort`, otherwise subsequent calls to `CreateJobObject` won't re-run the `Once.Do`, but also won't know the IOCP creation failed.